pandas/io/parsers.py

```diff
@@ -2123,9 +2123,27 @@ def _clean_mapping(mapping):
         else:
             clean_dtypes = _clean_mapping(self.dtype)
 
-        return self._convert_to_ndarrays(data, self.na_values, self.na_fvalues,
-                                         self.verbose, clean_conv,
-                                         clean_dtypes)
+        # Apply NA values.
+        clean_na_values = {}
+        clean_na_fvalues = {}
+
+        if isinstance(self.na_values, dict):
```

Comment:
Look at the `_get_na_list` method <a href="https://github.com/pandas-dev/pandas/blob/1725d24639a7c350a48fd201ca71b4548ea7186b/pandas/parser.pyx#L1260">here</a> for the C engine.  It gets called on every iteration to extract the relevant `na_values`.